### PR TITLE
[Dev] Add SiTU-GLU through Transformer Engine

### PR DESCRIPTION
**Pull request by harryzhou2000**
- [x] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

Add Kimi K3 SiTU-GLU as a global FFN activation on `dev`.

Paired main-target PR: https://github.com/NVIDIA/Megatron-LM/pull/6674

## Activation

For gate projection `G`, up projection `U`, and K3 defaults `beta1=4`, `beta2=25`:

```text
T_g = beta1 * tanh(G / beta1) * sigmoid(G)
T_u = beta2 * tanh(U / beta2)
Y   = T_g * T_u
```

Reference: [Kimi K3 technical report, Figure 4 and Equation 12](https://github.com/MoonshotAI/Kimi-K3/blob/main/k3_tech_report.pdf).

## Design

- Add `--situ-glu`, with `--moe-use-situ-glu` as an alias, and apply it consistently to dense, routed-expert, and shared-expert FFNs.
- Add `situ_glu_beta1=4` and `situ_glu_beta2=25` to `TransformerConfig`, with validation that SiTU-GLU is gated, unclamped, uses zero GLU linear offset, and has finite positive beta values.
- Add `situlu` as the correct PyTorch pointwise reference and configuration marker. It serves non-TE/custom module paths until PyTorch provides a dedicated `torch.nn.functional.situlu`-style operation; unary `F.silu` is not equivalent.
- Use `transformer_engine.pytorch.ops.SiTUGLU` for ordinary dense, sequential-expert, and ordinary shared-expert paths.
- Use `transformer_engine.pytorch.ops.ScaledSiTUGLU` for fused grouped dense, routed-expert, and shared-expert paths.
- Fail clearly when a requested TE path does not expose the required SiTU-GLU operator. cuDNN frontend capability selection remains inside TE.
- Extend checkpoint conversion, default FFN sizing, FLOP accounting, and memory estimation to treat SiTU-GLU as a gated GLU.

The required TE interface is available in https://github.com/NVIDIA/TransformerEngine/pull/3402.

With that TE interface, an older cuDNN frontend without SiTU parameters uses TE's unfused `GroupedLinear -> ScaledSiTUGLU -> GroupedLinear` operation sequence. A SiTU-capable cuDNN frontend enables fused grouped GEMM + SiTU-GLU without an MCore-side backend check.

## Test results

Validated on NVIDIA B300 with https://github.com/NVIDIA/TransformerEngine/pull/3402 and cuDNN Frontend develop through https://github.com/NVIDIA/cudnn-frontend/pull/670, using CuTe DSL 4.6.2:

- final focused SiTU-GLU and grouped-MLP suite on the review-fixed tip: **41 passed**;
- broader selected dev suites: **114 passed, 11 skipped, 4 deselected**; the deselected cases are existing multi-rank shared-expert parameterizations that cannot run on one GPU;
- BF16 real MCore MoE routed forward/backward through `ScaledSiTUGLU`: **passed**;
- MXFP8 real MCore MoE forward/backward: routed and fused-shared `ScaledSiTUGLU`, with the fused cuDNN/CuTe grouped-MLP operation asserted: **passed**;
- NVFP4 routed-expert forward/backward through `ScaledSiTUGLU`, with the fused Hadamard forward operation asserted: **passed**;
- `git diff --check`, Python compilation, and Ruff on all changed source and test files: **passed**;
- SSH commit signature and DCO: **passed**.
- `--use-checkpoint-args` restoration passed two parametrized SwiGLU/SiTU-GLU cases, including the SiTU selector and non-default beta values.
- The hybrid-model golden-config comparator passed on the current rebased tip.

Fused-shared NVFP4 currently reaches an existing TE FC1 weight-scale reshape failure that reproduces with an equivalent SwiGLU control, so this PR does not claim that combination. MXFP8 dSiTU validation used the cuDNN frontend correction included in https://github.com/NVIDIA/cudnn-frontend/pull/670; the published cuDNN frontend 1.27.0 package tested here does not yet contain that correction.

## Scope and dependencies

This PR depends on https://github.com/NVIDIA/TransformerEngine/pull/3402 for TE-backed SiTU-GLU functionality. The cuDNN frontend support is optional for correctness and required only for fused block-scaled grouped execution. A functional test is deferred until that TE interface is available in Megatron-LM CI; the focused MCore unit and GPU suites cover the current integration boundary.

This PR is independent of Quantile Balancing and https://github.com/NVIDIA/Megatron-LM/pull/6637.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: TODO — link a Megatron-LM feature-request issue before marking this draft ready for review.

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [x] I have added relevant documentation
- [x] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

Feel free to message or comment @NVIDIA/mcore-oncall to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

### CODE REVIEW

File: megatron/core/activations.py [hunk 1]
```diff
@@ -21,3 +21,17 @@ def quick_gelu(x: torch.Tensor) -> torch.Tensor:
...
+    """Apply SiTU-GLU to contiguous gate/up halves of an FC1 output.
+
+    This is the slow PyTorch reference and config marker until PyTorch provides
+    a dedicated ``torch.nn.functional.situlu``-style operation. Unary
+    ``F.silu`` is not equivalent because SiTU-GLU transforms both branches.
+    """
+    gate, up = torch.chunk(x, 2, dim=-1)
+    gate = beta1 * torch.tanh(gate / beta1) * torch.sigmoid(gate)
+    up = beta2 * torch.tanh(up / beta2)
+    return gate * up
```
Comment (yuzhongw-nvidia) [inline megatron/core/activations.py]:
```suggestion
    input_dtype = x.dtype
    gate, up = torch.chunk(x, 2, dim=-1)
    gate = gate.float()
    up = up.float()
    gate = beta1 * torch.tanh(gate / beta1) * torch.sigmoid(gate)
    up = beta2 * torch.tanh(up / beta2)
    return (gate * up).to(input_dtype)
```

Comment (yuzhongw-nvidia) [reply megatron/core/activations.py]: Kimi K3's released SituAndMul explicitly casts both gate and up halves to float32 before tanh/sigmoid, then casts only the product back to the input dtype (code). This fallback evaluates every operation in BF16/FP16 when given a low-precision FC1 output, so it is not release-equivalent.Please preserve the release precision contract here by upcasting both branches, computing the activation, and casting the result back, and add BF16 forward/backward coverage against the released formula; the current float64 self-equivalence test cannot detect this.

File: megatron/core/transformer/moe/experts.py [hunk 2]
```diff
@@ -487,14 +491,26 @@ def register_grouped_linear_params(
...
+            situ_glu_kwargs = {
+                "glu_interleave_size": glu_interleave,
+                "beta1": self.config.situ_glu_beta1,
+                "beta2": self.config.situ_glu_beta2,
+            }
+            if (
+                "activation_recompute_in_mlp"
+                in inspect.signature(te.pytorch.ops.ScaledSiTUGLU).parameters
+            ):
+                situ_glu_kwargs["activation_recompute_in_mlp"] = activation_recompute_in_mlp
```
Comment (yuzhongw-nvidia) [inline megatron/core/transformer/moe/experts.py]: `activation_recompute` becomes true for selective `moe_act` recompute, but the current Transformer Engine `_ScaledGLU` base explicitly raises for `activation_recompute_in_mlp=True`, including `ScaledSiTUGLU` (TE's own test asserts that rejection). `_is_fused_impl_supported()` does not reject this SiTU combination, so the path looks supported and then fails here during fused-op construction instead of falling back. The new fake-op test also accepts and asserts `True`, which is the opposite of the real TE contract. Please gate this combination and fall back, or require a TE capability that truly supports it, and update the test to model the actual class behavior.

File: megatron/core/activations.py (hunk 1 above)
Comment (harryzhou2000) [reply megatron/core/activations.py]: Fixed in `5770c6543`. The MCore fallback now casts both gate and up branches to FP32, evaluates the complete SiTU-GLU formula in FP32, and casts only the product back to the input dtype, matching Kimi K3's released `SituAndMul` precision contract. I also replaced the float64 self-equivalence check with BF16 forward/backward coverage against that FP32-compute reference. The focused dev suite passes (`115 passed, 11 skipped`), including the new BF16 test.

File: megatron/core/transformer/moe/experts.py (hunk 2 above)
Comment (harryzhou2000) [reply megatron/core/transformer/moe/experts.py]: Fixed in `5770c6543`. Signature presence is no longer treated as recompute capability: fused scaled-GLU selection now rejects selective `moe_act` recompute before op construction, and the SiTU path no longer forwards `activation_recompute_in_mlp` to `ScaledSiTUGLU`. The fake op and tests now model the real TE contract. I made the support guard generic because the current TE `_ScaledGLU` base rejects recompute for SwiGLU, SiTU-GLU, and clamped GLU; `ScaledSReLU` remains unaffected. The focused dev suite passes (`115 passed, 11 skipped`), and real TE was also checked to reject `ScaledSiTUGLU(activation_recompute_in_mlp=True)` as expected.

**Review by yuzhongw-nvidia — APPROVED** [review]
LGTM. Thanks for your outstanding work!

**Comment by yuzhongw-nvidia** [comment]
/ok to test https://github.com/NVIDIA/Megatron-LM/commit/4edf7500c76fc0eb08bc107c02800a8092c2179b

File: megatron/core/transformer/moe/experts.py [hunk 3]
```diff
@@ -384,7 +385,14 @@ def _activation_name():
                 f"(gated_linear_unit={self.config.gated_linear_unit}, "
                 f"use_fused_weighted_squared_relu={self.config.use_fused_weighted_squared_relu})"
             )
-        if self.config.activation_func == F.silu:
+        if use_glu_fusion and self.activation_recompute:
+            return _unsupported(
+                "Transformer Engine scaled GLUs do not support activation recompute"
+            )
```
[the remark this answers is not in the document]
Comment (harryzhou2000) [reply megatron/core/transformer/moe/experts.py:440]: Fixed in `74aafeb52`. The recompute guard is now scoped only to `situlu`/`ScaledSiTUGLU`. Existing SwiGLU and quick-GEGLU paths retain their prior activation-recompute plumbing, with a parametrized regression test that sets `activation_recompute=True` for both. The final focused suites pass on dev (41 tests) and main (60 tests).

File: megatron/core/activations.py [hunk 4]
```diff
@@ -21,3 +21,20 @@ def quick_gelu(x: torch.Tensor) -> torch.Tensor:
 def fast_gelu(x: torch.Tensor) -> torch.Tensor:
     """Fast GELU activation"""
     return 0.5 * x * (1.0 + torch.tanh(x * 0.7978845608 * (1.0 + 0.044715 * x * x)))
+
+
+@jit_fuser
+def situlu(x: torch.Tensor, beta1: float = 4.0, beta2: float = 25.0) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (harryzhou2000) [reply megatron/core/activations.py:27]: Confirmed. `jit_fuser` decorates `situlu` once, and toggling the module-global fuser does not rebind that callable, so the identity-based dispatch remains stable. The beta values are configuration invariants: each call site reuses the same pair for the lifetime of a config/module; a different config may compile a separate specialization, but the values do not vary per invocation. Non-default betas remain covered by the configuration/op-selection tests, so no code change was needed for this observation.

**Comment by harryzhou2000** [comment]
Addressed the current review findings on both SiTU-GLU PRs in `74aafeb52` (dev) and `89b265c25` (main):

- Scoped the activation-recompute rejection to `ScaledSiTUGLU` only, with regressions proving existing SwiGLU and quick-GEGLU recompute support is unchanged.
- Marked checkpoint-converter FC1 metadata as gated for either SwiGLU or SiTU-GLU.
- Added independent fixed-value coverage, including positive and negative saturation.
- Applied the repository's Black 26.3.0/isort/Ruff formatting checks.

I kept the CLI's TE selection intentional: stock SiTU-GLU paths require the TE operators, while `situlu` remains the numerically correct pointwise reference/configuration marker for direct/custom use. The pre-existing `TEActivationOp` ReGLU/ReLU typo is unrelated and left out of this focused change. A functional case is noted as deferred until https://github.com/NVIDIA/TransformerEngine/pull/3402 is available in Megatron-LM CI.

Final focused GPU suites on NVIDIA B300: dev **41 passed**; main **60 passed**.

File: megatron/training/checkpointing.py [hunk 5]
```diff
@@ -1369,7 +1369,7 @@ def generate_state_dict(
 def preprocess_fsdp_dtensor_state_dict(args, raw_state_dict, model):
     state_dict = raw_state_dict.copy()
     handle_fp8_extra_state_case(state_dict["model"])
-    if args.swiglu:
+    if args.swiglu or getattr(args, "situ_glu", False):
```
Comment (FDecaYed) [inline megatron/training/checkpointing.py:1372]: `--use-checkpoint-args` does not currently restore `situ_glu`. `load_args_from_checkpoint()` restores `squared_relu` and `swiglu`, but never calls `_set_arg("situ_glu", force=True)`. As a result, loading a SiTU-GLU checkpoint with `--use-checkpoint-args` reconstructs it as a non-gated MLP unless the user also repeats `--situ-glu`, which leads to an FC1 shape mismatch. Could we restore `situ_glu` from the checkpoint args as well?

Comment (harryzhou2000) [reply megatron/training/checkpointing.py:1372]: Fixed in `5c1d0f6e2` and mirrored in the main-target PR at `c7a03fa7d`. `load_args_from_checkpoint()` now restores `situ_glu` immediately beside `swiglu`, and also restores `situ_glu_beta1` / `situ_glu_beta2` so non-default checkpoints keep their numerical contract. A parametrized regression test covers both SwiGLU and SiTU-GLU restoration (2 cases passed), and the hybrid-model golden-config comparator also passes on both rebased tips.

File: megatron/core/transformer/moe/experts.py [hunk 6]
```diff
@@ -384,7 +385,14 @@ def _activation_name():
                 f"(gated_linear_unit={self.config.gated_linear_unit}, "
                 f"use_fused_weighted_squared_relu={self.config.use_fused_weighted_squared_relu})"
             )
-        if self.config.activation_func == F.silu:
+        if self.config.activation_func is situlu and self.activation_recompute:
```
Comment (FDecaYed) [inline megatron/core/transformer/moe/experts.py]: TE #3402, which this PR depends on, rejects `activation_recompute_in_mlp=True` for all scaled GLUs (`ScaledSwiGLU`, `ScaledSiTUGLU`, and `ScaledClampedQGeGLU`), not just SiTU. With this guard scoped to `situlu`, SwiGLU and quick-GEGLU still pass `_is_fused_impl_supported()`, then `_make_fused_ops()` forwards `True` and TE raises during op construction. The current regression test uses fake ops that accept this value, so it does not model the required TE dependency. Could we restore the generic fallback for scaled GLUs when `activation_recompute` is enabled and update the test to match TE #3402? `ScaledSReLU` can remain supported.

**Review by FDecaYed — CHANGES_REQUESTED** [review]
Requesting changes for the recompute compatibility issue below.

**Review by FDecaYed — APPROVED** [review]
Re-reviewed the current head after the scaled-GLU recompute fallback fix. The TE contract is now handled consistently, including clamped SwiGLU and quick-GEGLU, while ScaledSReLU remains supported.

**Comment by FDecaYed** [comment]
dev has moved since this head, and the PR is now `CONFLICTING/DIRTY`. Could you rebase `hhanyu/situ-glu-te-dev` onto the current `dev` and preserve the existing SiTU-GLU fixes? The previous exact-head CI is green, but it no longer qualifies the rebased head. Once the new SHA is up, I can re-review it and restart functional CI.
